In [1]:
%%javascript
/**********************************************************************************************
Known Mathjax Issue with Chrome - a rounding issue adds a border to the right of mathjax markup
https://github.com/mathjax/MathJax/issues/1300
A quick hack to fix this based on stackoverflow discussions: 
http://stackoverflow.com/questions/34277967/chrome-rendering-mathjax-equations-with-a-trailing-vertical-line
**********************************************************************************************/

$('.math>span').css("border-left-color","transparent")

<IPython.core.display.Javascript object>

In [2]:
%reload_ext autoreload
%autoreload 2

# DAMLAS - Machine Learning At Scale
## Assignment - HW4
Data Analytics and Machine Learning at Scale
Target, Minneapolis

---
__Name:__  Tom Hughes   
__Class:__ DAMLAS (Section Summer 2016)     
__Email:__  Thomas.Hughes@Target.com     
__Week:__   04

# Table of Contents <a name="TOC"></a> 

1.  [HW Introduction](#1)   
2.  [HW References](#2)
3.  [HW 4 Problems](#3)   
    4.0.  [Final Project description](#4.0)   
    4.1.  [Build a decision to predict whether you can play tennis or no](#4.1)   
    4.2.  [Regression Tree (OPTIONAL Homework)](#4.2)    
    4.3.  [Predict survival on the Titanic](#4.3)    
    4.4.  [Heritage Healthcare Prize (Predict # Days in Hospital next year)](#4.4)  


<a name="1">
# 1 Instructions
[Back to Table of Contents](#TOC)
* Homework submissions are due by Thursday, 08/18/2016 at 11AM (CT).


* Prepare a single Jupyter notebook (not a requirment), please include questions, and question numbers in the questions and in the responses.
Submit your homework notebook via the following form:

   + [Submission Link - Google Form](http://goo.gl/forms/er3OFr5eCMWDngB72)


### Documents:
* IPython Notebook, published and viewable online.
* PDF export of IPython Notebook.

<a name="2">
# 2 Useful References
[Back to Table of Contents](#TOC)

* [Lecture Slides on Decision Trees and Ensembles](https://www.dropbox.com/s/lm4vuocqoq6mq7k/Lecture-13-Decision-Trees-PLanet.pdf?dl=0)

* Chapter 17 on decision Trees,   https://www.dropbox.com/s/5ca98ah5chqlcmn/Data_Science_from_Scratch%20%281%29.pdf?dl=0   [Please do not share this PDF]
* Karau, Holden, Konwinski, Andy, Wendell, Patrick, & Zaharia, Matei. (2015). Learning Spark: Lightning-fast big data analysis. Sebastopol, CA: O’Reilly Publishers.
* Hastie, Trevor, Tibshirani, Robert, & Friedman, Jerome. (2009). The elements of statistical learning: Data mining, inference, and prediction (2nd ed.). Stanford, CA: Springer Science+Business Media. __(Download for free [here](http://statweb.stanford.edu/~tibs/ElemStatLearn/printings/ESLII_print10.pdf))__
* Ryza, Sandy, Laserson, Uri, Owen, Sean, & Wills, Josh. (2015). Advanced analytics with Spark: Patterns for learning from data at scale. Sebastopol, CA: O’Reilly Publishers.
---

---

## 3.  HW4  <a name="4"></a>
[Back to Table of Contents](#TOC)

 <a name="4.0"></a>
## HW4.0 Final Project description

Please prepare your project description using the following format
* 200 words abstract
* data source and description
* pipeline of steps (in a block diagram)
* Metrics for success

PLEASE NOTE: We will probably have project team sizes of 3 people plus/minus 1

### Project Description

#### Team:  Mark VonOven, Earl Sun, Thomas Hughes

__Situation:__   We have a record of what guests browse online.  There is no such analog for our Stores.  We know what guests buy, but do not know how they shopped.  There is value in understanding the path guests take, what material can be presented along the way and where to efficiently place our team.

__Task:__  LED lights are being installed in stores which are capable of tracking guests with the Target app.  Using position information we can plot the guests' path, items they viewed and where they spent the most time.  This will allow us to understand and predict how a guest's path impacts purchases.

__Data Source:__  Position data is stored within Target's hadoop environment.  The information contains x/y coordinates, yaw and a timestamp.  We will also need data on Store layout and purchases made.

__Pipeline of Steps:__
1.  Obtain position data, Store layout and purchase record
2.  Analyze path, speed, stops and purchases for TMs and guests
3.  Approximate location when accuracy switches from LED to Bluetooth
4.  Predict guest path based on prior history and optimize tailored marketing

__Success Metrics:__  
1. Identify how Store path translates to sales
2. Determine whether team member placement impacts sales

 <a name="4.1"></a>
## HW4.1 Build a decision to predict whether you can play tennis or not

[Back to Table of Contents](#TOC)

Decision Trees

Write a program in Python (or in Spark; this part is optional) to implement the ID3 decision tree algorithm. You should build a tree to predict PlayTennis, based on the other attributes (but, do not use the Day attribute in your tree.). You should read in a space delimited dataset in a file called dataset.txt and output to the screen your decision tree and the training set accuracy in some readable format. For example, here is the tennis dataset. The first line will contain the names of the fields:

<PRE>
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no
d8 sunny cool normal FALSE yes
d9 rainy mild normal FALSE yes
d10 sunny mild normal TRUE yes
d11 overcast mild high TRUE yes
d12 overcast hot normal FALSE yes
d12 rainy mild high TRUE no
</PRE>

The last column is the classification attribute, and will always contain contain the values yes or no.

For output, you can choose how to draw the tree so long as it is clear what the tree is. You might find it easier if you turn the decision tree on its side, and use indentation to show levels of the tree as it grows from the left. For example:

<PRE>
outlook = sunny
|  humidity = high: no
|  humidity = normal: yes
outlook = overcast: yes
outlook = rainy
|  windy = TRUE: no
|  windy = FALSE: yes

</PRE>

You don't need to make your tree output look exactly like above: feel free to print out something similarly readable if you think it is easier to code.

You may find Python dictionaries especially useful here, as they will give you a quick an easy way to help manage counting the number of times you see a particular attribute.

Here are some FAQs that I've gotten in the past regarding this assignment, and some I might get if I don't answer them now.

__Should my code work for other datasets besides the tennis dataset?__ 
Yes. We will give your program a different dataset to try it out with. You may assume that our dataset is correct and well-formatted, but you should not make assumptions regrading number of rows, number of columns, or values that will appear within. The last column will also be the classification, and will always contain yes or no values.

__Is it possible that some value, like "normal," could appear in more than one column?__
Yes. In addition to the column "humidity", we might have had another column called "skycolor" which could have values "normal," "weird," and "bizarre."

__Could "yes" and "no" appear as possible values in columns other than the classification column?__
Yes. In addition to the classification column "playtennis," we might have had another column called "seasonalweather" which would contain "yes" and "no."

In [48]:
%%writefile tennisData.txt
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no
d8 sunny cool normal FALSE yes
d9 rainy mild normal FALSE yes
d10 sunny mild normal TRUE yes
d11 overcast mild high TRUE yes
d12 overcast hot normal FALSE yes
d12 rainy mild high TRUE no

Overwriting tennisData.txt


In [49]:
import math
from collections import Counter
from collections import defaultdict
from functools import partial



def entropy(class_probabilities):
    """given a list of class probabilities, compute the entropy"""
    return sum(-p * math.log(p, 2)
        for p in class_probabilities
        if p) # ignore zero probabilities
        
def class_probabilities(labels):
    total_count = len(labels)*1.0
    return [count / total_count
        for count in Counter(labels).values()]  

    
def data_entropy(labeled_data):
    labels = [label for _, label in labeled_data]
    probabilities = class_probabilities(labels)
    return entropy(probabilities)


def partition_entropy(subsets):
    """find the entropy from this partition of data into subsets
    subsets is a list of lists of labeled data"""
    total_count = sum(len(subset) for subset in subsets)
    return sum( data_entropy(subset) * len(subset) / total_count
                for subset in subsets )
                
def partition_by(inputs, attribute):
    """each input is a pair (attribute_dict, label).
    returns a dict : attribute_value -> inputs"""
    groups = defaultdict(list)
    for input in inputs:
        key = input[0][attribute] # get the value of the specified attribute
        groups[key].append(input) # then add this input to the correct list
    return groups
    
def partition_entropy_by(inputs, attribute):
    """computes the entropy corresponding to the given partition"""
    partitions = partition_by(inputs, attribute)
    return partition_entropy(partitions.values()) 
        

def build_tree_id3(inputs, split_candidates=None):
    # if this is our first pass,
    # all keys of the first input are split candidates

    if split_candidates is None:
        split_candidates = inputs[0][0].keys()
    # count Trues and Falses in the inputs
    num_inputs = len(inputs)
    num_trues = len([label for item, label in inputs if label])
    num_falses = num_inputs - num_trues
    if num_trues == 0: return False # no Trues? return a "False" leaf
    if num_falses == 0: return True # no Falses? return a "True" leaf
    
    if not split_candidates: # if no split candidates left
        return num_trues >= num_falses # return the majority leaf
    
    # otherwise, split on the best attribute
    best_attribute = min(split_candidates,
                             key=partial(partition_entropy_by, inputs))
    
    partitions = partition_by(inputs, best_attribute)
    new_candidates = [a for a in split_candidates
                    if a != best_attribute]
    
    # recursively build the subtrees
    subtrees = { attribute_value : build_tree_id3(subset, new_candidates)
                for attribute_value, subset in partitions.iteritems() }
    
    subtrees[None] = num_trues > num_falses # default case
    
    return (best_attribute, subtrees)  
    
    
def classify(tree, input):
    """classify the input using the given decision tree"""
    
    # if this is a leaf node, return its value
    if tree in [True, False]:
        return tree
    
    # otherwise this tree consists of an attribute to split on
    # and a dictionary whose keys are values of that attribute
    # and whose values of are subtrees to consider next
    attribute, subtree_dict = tree
    
    subtree_key = input.get(attribute) # None if input is missing attribute
    
    if subtree_key not in subtree_dict: # if no subtree for key,
        subtree_key = None # we'll use the None subtree
    
    subtree = subtree_dict[subtree_key] # choose the appropriate subtree
    return classify(subtree, input) # and use it to classify the input    
    


def print_tree(tree):
    print tree[0]
    for key in tree[1].keys():
        if key != None:
            if tree[1][key] == True:
                print "\n"+" | "+str(key)
                print "    "+str(tree[1][key])
            elif tree[1][key] == False:
                print "\n"+" | "+str(key)
                print "    "+str(tree[1][key])
            else:
                print "\n"+" | "+str(key)
                for k in tree[1][key][1].keys():
                    if k != None:
                            print "   |" + str(tree[1][key][0])+" = "+str(k)
                            print "    "+str(tree[1][key][1][k])
                            


In [50]:
filenm = "tennisData.txt"

def inputformat(data):
    headers = []
    input_data = []
    with open(data) as f:
        headers = f.readline().split(" ") 
        headers.pop(0)
        for line in f.readlines():
            d = line.strip().split(" ")
            d.pop(0)
            rng = len(d)-1
            input_line = {}
            if d[rng] == 'yes':
                d[rng] = True
            else:
                d[rng] = False
            for i in range(rng):
                input_line[headers[i]] = d[i]
            input_data.append((input_line, d[rng]))  
    return input_data            
    
inputs2 = inputformat(filenm)
        
tree = build_tree_id3(inputs2)
print_tree(tree)

outlook

 | rainy
   |wind = FALSE
    True
   |wind = TRUE
    False

 | overcast
    True

 | sunny
   |humidity = high
    False
   |humidity = normal
    True


__ HW4.1.1 What is the classification accuracy of the tree on the training data?__


In [61]:
def fcst_acc(data):
    accuracy = 0
    total = len(data)
    for i in data:
        if i[1] == classify(tree, i[0]):
            accuracy += 1    

    print "accuracy is: "+str((1.0*accuracy)/(1.0*total)*100)+"%"



fcst_acc(inputs2)
    


accuracy is: 100.0%


__HW4.1.2  Is it possible to produce some set of correct training examples that will get the algorihtm
to include the attribute Temperature in the learned tree, even though the true target concept is
independent of Temperature? if no, explain. If yes, give such a set. __

In [57]:
%%writefile tennisData_overfit.txt
Day outlook temperature humidity wind playtennis
d1 sunny cool high FALSE no
d2 sunny cool high TRUE no
d3 overcast hot high FALSE yes
d4 rainy hot high FALSE yes
d5 rainy hot normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast hot normal TRUE yes
d7 sunny cool high FALSE no
d8 sunny hot normal FALSE yes
d9 rainy hot normal FALSE yes
d10 sunny hot normal TRUE yes
d11 overcast hot high TRUE yes
d12 overcast hot normal FALSE yes
d12 rainy cool high TRUE no

Overwriting tennisData_overfit.txt


### The set above has the target concept output by ID3 in 4.1.  By changing the temperature values all to 'hot' when playtennis = 'yes' and 'cool' when playtennis = 'no', ID3 would split on that factor first, even though the true concept is not related to temperature.  That is the drawback of ID3.  Since it constantly searches for the optimal entropy, it can often lead to overfitting the data.  This is why random forest and boosted trees use techniques like only training on a subset of columns. 

__HW4.1.3  Now, build a tree using only examples D1–D7. What is the classification accuracy for the
training set? what is the accuracy for the test set (examples D8–D14)? explain why you think these
are the results.__

In [47]:
%%writefile tennisData_sample.txt
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no

Overwriting tennisData_sample.txt


In [58]:
filenm = "tennisData_sample.txt"

def inputformat(data):
    headers = []
    input_data = []
    with open(data) as f:
        headers = f.readline().split(" ") 
        headers.pop(0)
        for line in f.readlines():
            d = line.strip().split(" ")
            d.pop(0)
            rng = len(d)-1
            input_line = {}
            if d[rng] == 'yes':
                d[rng] = True
            else:
                d[rng] = False
            for i in range(rng):
                input_line[headers[i]] = d[i]
            input_data.append((input_line, d[rng]))  
    return input_data            
    
inputs2 = inputformat(filenm)
        
tree = build_tree_id3(inputs2)
print_tree(tree)

outlook

 | rainy
   |wind = FALSE
    True
   |wind = TRUE
    False

 | overcast
    True

 | sunny
    False


In [59]:
%%writefile tennisData_test.txt
Day outlook temperature humidity wind playtennis
d8 sunny cool normal FALSE yes
d9 rainy mild normal FALSE yes
d10 sunny mild normal TRUE yes
d11 overcast mild high TRUE yes
d12 overcast hot normal FALSE yes
d12 rainy mild high TRUE no

Overwriting tennisData_test.txt


In [62]:
filenm = "tennisData_test.txt"

def inputformat(data):
    headers = []
    input_data = []
    with open(data) as f:
        headers = f.readline().split(" ") 
        headers.pop(0)
        for line in f.readlines():
            d = line.strip().split(" ")
            d.pop(0)
            rng = len(d)-1
            input_line = {}
            if d[rng] == 'yes':
                d[rng] = True
            else:
                d[rng] = False
            for i in range(rng):
                input_line[headers[i]] = d[i]
            input_data.append((input_line, d[rng]))  
    return input_data            
    
inputs2 = inputformat(filenm)
fcst_acc(inputs2)

accuracy is: 66.6666666667%


### Accuracy is 66.67%.  This is because the tree is overfitting to the training examples

__HW4.1.4 In this case, and others, there are only a few labelled examples available for training (that
is, no additional data is available for testing or validation). Suggest a concrete pruning strategy, that
can be readily embedded in the algorithm, to avoid over fitting. Explain why you think this strategy
should work.__

A pruning strategy that will work for ID3 is sub-tree replacement pruning.  First, take the training data and split it randomly into a training set and a test set (you could do k-folds cross validation for this to make it more robust).  Next, grow the tree for the training set and use it to classify the test set.  After classification, remove the lowest leaves from the tree and measure the new error on the validation set.  Remove the node with the greatest error improvement.  Repeat.  This strategy works by forcing a tree from it's "pure" state to be more generalized and thus more effective on new/unseen data.

 <a name="4.2"></a>
 ## HW4.2 Regression Tree (OPTIONAL Homework) 
 
[Back to Table of Contents](#TOC)

Implement a decision tree algorithm for regression for two input continous variables and one categorical input variable on a single core computer using Python. 

- Use the IRIS dataset to evaluate your code, where the input variables are: Petal.Length Petal.Width  Species  and the target or output variable is  Sepal.Length. 
- Use the same dataset to train and test your implementation. 
- Stop expanding nodes once you have less than ten (10) examples (along with the usual stopping criteria). 
- Report the mean squared error for your implementation and contrast that with the MSE from scikit-learn's implementation on this dataset (http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html)


 <a name="4.3"></a>
## HW4.3 Predict survival on the Titanic using Python (Logistic regression, SVMs, Random Forests)

[Back to Table of Contents](#TOC)

The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

In this challenge, you need to review (and edit the code) in this [notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/kmbgrkhh73931lo/Titanic-EDA-LogisticRegression.ipynb) to do analysis of what sorts of people were likely to survive. In particular, please look at how the tools of machine learning are used to predict which passengers survived the tragedy. Please share any usefule graphs/analysis you come up with via the group email.

For more details see:

* https://www.kaggle.com/c/titanic

### Here's a link to my Google Group email:  https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/target-msp-ds-camp-2016-06/-3wB2HNW-p4


 <a name="4.4"></a>
 ## HW4.4 Heritage Healthcare Prize (Predict # Days in Hospital next year)
[Back to Table of Contents](#TOC)

1. Introduction 
Back to Table of Contents

The Heritage Health Prize (HHP) was a data science challenge sponsored by The Heritage Provider Network. It took place from April 4, 2011 to April 4, 2013. For information on the winning entries, please see here.

Please see the following notebooks for more background and candidate solutions


- Spark Map-Reduce + MMLlib solution (with optional extensions) See [Notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/v52cxipe7yftf97/HeritageHealthPrizeUnitTestNotebook_Spark-Map-Reduce.ipynb)

- Spark SQL + MLLib solution (with optional extensions): [Notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/s2wxg6g982oho5m/HeritageHealthPrizeUnitTestNotebook_SQL_FINAL.ipynb)


Please look at section 7 in both notebooks complete any one or more the suggested next steps. E.g.,

* Please complete the EDA extensions using inspiration from the Titanic Notebook from above.
* __Complete Section 3.B: EDA-0. Gather information to see what transformations may need to be done on the data.__
Answer questions about each raw DataFrame. In general, is the data in good shape? For example, in each of the Target DataFrames (df_target_Y1, df_target_Y2, df_target_Y3), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers? In the Claims DataFrame (df_claims), how many different ProviderIDs are there? How many different PrimaryConditionGroups are there? What are their values? What values can the CharlesonIndex take on? Are they integers? In the Drug Count DataFrame (df_drug_count), what values can DrugCount take on? Are they all integers? Given this information, what transformations are needed?

* __Complete Section 3.D: EDA-1. Create tables and graphs to display information about the transformed DataFrames. __
For inspiration, see the Titanic notebook discussed above. Answer questions about each DataFrame. For example, in each of the Target DataFrames (df_target_Y1, df_target_Y2, df_target_Y3), what is the minimum, maximum, mean, and standard deviation of DaysInHospital? In the Claims DataFrame, group by MemberID and Year and count the number of records. What is the minimum, maximum, mean, and standard deviation of the count? Do the same for the Drug Count and Lab Count DataFrames, etc.


* __ Please generate ensemble of DT model using 100 trees with 8 nodes and report the Loss __
Try additional models. See possibilities here (e.g. Decision Tree Regressor, Gradient-Boosted Trees Regressor, Random Forest Regressor). See an example here. Tune their hyperparameters. Try different feature selections. Try a two-step model.


# 4.4_Hughes <a name="Hughes"></a>

[Back to Table of Contents](#TOC)

[Claims Analysis](#4.4Claims)

[Drug Analysis](#4.4Drug)

[Lab Analysis](#4.4Lab)

[Member Analysis](#4.4Member)

[Target_Yr2 Analysis](#4.4Y2)

[Target_Yr3 Analysis](#4.4Y3)

[Target_Yr4 Analysis](#4.4Y4)

* __Complete Section 3.B: EDA-0. Gather information to see what transformations may need to be done on the data.__
Answer questions about each raw DataFrame. In general, is the data in good shape? For example, in each of the Target DataFrames (df_target_Y1, df_target_Y2, df_target_Y3), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers? In the Claims DataFrame (df_claims), how many different ProviderIDs are there? How many different PrimaryConditionGroups are there? What are their values? What values can the CharlesonIndex take on? Are they integers? In the Drug Count DataFrame (df_drug_count), what values can DrugCount take on? Are they all integers? Given this information, what transformations are needed?

In [3]:
import os
import sys #current as of 7/15/2016


spark_home = os.environ['SPARK_HOME'] = 'c:/Apps/Spark/Spark_1.6.2/'
if not spark_home:
    raise ValueError('SPARK_HOME enviroment variable is not set')
sys.path.insert(0,os.path.join(spark_home,'python'))
sys.path.insert(0,os.path.join(spark_home,'python/lib/py4j-0.9-src.zip'))

# First, we initialize the Spark environment

#import findspark
#findspark.init()

import pyspark
from pyspark.sql import SQLContext

# We can give a name to our app (to find it in Spark WebUI) and configure execution mode
# In this case, it is local multicore execution with "local[*]"
app_name = "example-logs"
master = "local[*]"
conf = pyspark.SparkConf().setAppName(app_name).setMaster(master)
sc = pyspark.SparkContext(conf=conf)
sqlContext = SQLContext(sc)


print(sc)
print(sqlContext)


# Import some libraries to work with dates
import dateutil.parser
import dateutil.relativedelta as dateutil_rd

In [4]:
import os

# Set the directory containing the raw data files
DATA_DIR = os.path.join('HealthcareData')

In [5]:
import pandas as pd
from pyspark.sql.types import *
import pyspark.sql.functions as func

In [4]:

def load_csv_file(filename):
    # HACK - Loading csv directly into a dataframe was supposed to work by using a Databricks
    # package (https://github.com/databricks/spark-csv), but it didn't.
    # So in this section, we are going the Pandas -> DataFrame route.  Could also go RDD -> DataFrame route.
    
    # Set our file path
    input_path = os.path.join(DATA_DIR, filename)
    
    # Read into Pandas dataframe.  We assume the file contains a header row.
    # Note:  We read in all fields as string.
    df_pandas = pd.read_csv(input_path, dtype='str')  
    
    # Read Pandas dataframe into SQL DataFrame.
    # We must define the schema so that datatype is not automatically inferred and everything remains string.
    schema_string = df_pandas.columns.values.tolist()
    fields = [StructField(field_name, StringType(), True) for field_name in schema_string]
    schema = StructType(fields)
    
    return sqlContext.createDataFrame(df_pandas, schema)

# Read in Year 2 Target Variables
df_target_Y2 = load_csv_file('DaysInHospital_Y2.csv').cache()
# Check the data
print "Schema for target_Y2 DataFrame with %d rows:" %(df_target_Y2.count())
df_target_Y2.printSchema()
df_target_Y2.show(1)

# Read in Year 3 Target Variables
df_target_Y3 = load_csv_file('DaysInHospital_Y3.csv').cache()
# Check the data
print "Schema for target_Y3 DataFrame with %d rows:" %(df_target_Y3.count())
df_target_Y3.printSchema()
df_target_Y3.show(1)

# Read in Year 4 Target Variables
df_target_Y4 = load_csv_file('Target.csv').cache()
# Check the data
print "Schema for target_Y4 DataFrame with %d rows:" %(df_target_Y4.count())
df_target_Y4.printSchema()
df_target_Y4.show(1)

# Read in Claims Data
df_claims = load_csv_file('Claims.csv').cache()
# Check the data
print "Schema for claims DataFrame with %d rows:" %(df_claims.count())
df_claims.printSchema()
df_claims.show(1)

# Read in Drug Data
df_drug_count = load_csv_file('DrugCount.csv').cache()
# Check the data
print "Schema for drug_count DataFrame with %d rows:" %(df_drug_count.count())
df_drug_count.printSchema()
df_drug_count.show(1)

# Read in Lab Data
df_lab_count = load_csv_file('LabCount.csv').cache()
# Check the data
print "Schema for lab_count DataFrame with %d rows:" %(df_lab_count.count())
df_lab_count.printSchema()
df_lab_count.show(1)

# Read in Members Data
df_members = load_csv_file('Members.csv').cache()
# Check the data
print "Schema for members DataFrame with %d rows:" %(df_members.count())
df_members.printSchema()
df_members.show(1)

Schema for target_Y2 DataFrame with 76038 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospital: string (nullable = true)

+--------+---------------+--------------+
|MemberID|ClaimsTruncated|DaysInHospital|
+--------+---------------+--------------+
|24027423|              0|             0|
+--------+---------------+--------------+
only showing top 1 row

Schema for target_Y3 DataFrame with 71435 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospital: string (nullable = true)

+--------+---------------+--------------+
|MemberID|ClaimsTruncated|DaysInHospital|
+--------+---------------+--------------+
|90963501|              0|             0|
+--------+---------------+--------------+
only showing top 1 row

Schema for target_Y4 DataFrame with 70942 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospita

In [6]:
print sqlContext

In [6]:
def load_csv_file(filename):
    # HACK - Loading csv directly into a dataframe was supposed to work by using a Databricks
    # package (https://github.com/databricks/spark-csv), but it didn't.
    # So in this section, we are going the Pandas -> DataFrame route.  Could also go RDD -> DataFrame route.
    
    # Set our file path
    input_path = os.path.join(DATA_DIR, filename)
    
    # Read into Pandas dataframe.  We assume the file contains a header row.
    # Note:  We read in all fields as string.
    df_pandas = pd.read_csv(input_path, dtype='str')  
    
    # Read Pandas dataframe into SQL DataFrame.
    # We must define the schema so that datatype is not automatically inferred and everything remains string.
    schema_string = df_pandas.columns.values.tolist()
    fields = [StructField(field_name, StringType(), True) for field_name in schema_string]
    schema = StructType(fields)
    
    return sqlContext.createDataFrame(df_pandas, schema)

## Claims Data <a name="4.4Claims"></a>

[Back to 4.4](#Hughes)

In the Claims DataFrame (df_claims), how many different ProviderIDs are there? How many different PrimaryConditionGroups are there? What are their values? What values can the CharlesonIndex take on? Are they integers? 

In [11]:
# Read in Claims Data
df_claims = load_csv_file('Claims.csv').cache()
# Check the data
print "Schema for claims DataFrame with %d rows:" %(df_claims.count())
df_claims.printSchema()
df_claims.show(1)  

Schema for claims DataFrame with 2668990 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ProviderID: string (nullable = true)
 |-- Vendor: string (nullable = true)
 |-- PCP: string (nullable = true)
 |-- Year: string (nullable = true)
 |-- Specialty: string (nullable = true)
 |-- PlaceSvc: string (nullable = true)
 |-- PayDelay: string (nullable = true)
 |-- LengthOfStay: string (nullable = true)
 |-- DSFS: string (nullable = true)
 |-- PrimaryConditionGroup: string (nullable = true)
 |-- CharlsonIndex: string (nullable = true)
 |-- ProcedureGroup: string (nullable = true)
 |-- SupLOS: string (nullable = true)

+--------+----------+------+-----+----+---------+--------+--------+------------+-----------+---------------------+-------------+--------------+------+
|MemberID|ProviderID|Vendor|  PCP|Year|Specialty|PlaceSvc|PayDelay|LengthOfStay|       DSFS|PrimaryConditionGroup|CharlsonIndex|ProcedureGroup|SupLOS|
+--------+----------+------+-----+----+---------+--------+--------+----

In [13]:
# Method 2: Direct queries of DataFrames
# DataFrame Operations:  https://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.DataFrame

# Example: Get counts of values in each column of Claims data

 


print df_claims.columns
columns_for_analysis = ['Year', 'Specialty', 'PlaceSvc', 'LengthOfStay', 'DSFS', 
                        'PrimaryConditionGroup', 'CharlsonIndex', 'ProcedureGroup', 'SupLOS']



for column in columns_for_analysis:
    df_claims.groupBy(column).count().show()
       

['MemberID', 'ProviderID', 'Vendor', 'PCP', 'Year', 'Specialty', 'PlaceSvc', 'PayDelay', 'LengthOfStay', 'DSFS', 'PrimaryConditionGroup', 'CharlsonIndex', 'ProcedureGroup', 'SupLOS']
+----+------+
|Year| count|
+----+------+
|  Y1|865689|
|  Y2|898872|
|  Y3|904429|
+----+------+

+--------------------+------+
|           Specialty| count|
+--------------------+------+
|                 NaN|  8405|
|Obstetrics and Gy...| 36594|
|               Other| 92687|
|  Diagnostic Imaging|207297|
|          Pediatrics| 84862|
|      Rehabilitation| 57554|
|            Internal|672059|
|           Emergency|126130|
|             Surgery|208217|
|          Laboratory|653188|
|      Anesthesiology| 33435|
|    General Practice|473655|
|           Pathology| 14907|
+--------------------+------+

+-------------------+-------+
|           PlaceSvc|  count|
+-------------------+-------+
|                NaN|   7632|
|        Urgent Care| 199528|
|              Other|  11700|
|    Independent Lab| 65775

In [7]:
df_claims = load_csv_file('Claims.csv').cache()
df_claims.dtypes

[('MemberID', 'string'),
 ('ProviderID', 'string'),
 ('Vendor', 'string'),
 ('PCP', 'string'),
 ('Year', 'string'),
 ('Specialty', 'string'),
 ('PlaceSvc', 'string'),
 ('PayDelay', 'string'),
 ('LengthOfStay', 'string'),
 ('DSFS', 'string'),
 ('PrimaryConditionGroup', 'string'),
 ('CharlsonIndex', 'string'),
 ('ProcedureGroup', 'string'),
 ('SupLOS', 'string')]

In [8]:
df_claims.groupby('ProviderID').count().distinct().show()

+----------+-----+
|ProviderID|count|
+----------+-----+
|    944326|  208|
|   4028868| 1953|
|   2191832| 1703|
|    455068|   13|
|   7202236|   17|
|   2147967|  314|
|   4112611|  212|
|   6168173|   41|
|   1192364|   45|
|   3584196|  107|
|   8840994|   31|
|   8843667|   40|
|   9376562|    9|
|   3712896|    9|
|   9388307|    4|
|    671509|    3|
|   4910521|    4|
|   4547282|    6|
|   2864312|   14|
|    892252|   21|
+----------+-----+
only showing top 20 rows



In [22]:
df_claims.agg(func.countDistinct('ProviderID')).show()

+-----------------+
|count(ProviderID)|
+-----------------+
|            14700|
+-----------------+



In [26]:
df_claims.agg(func.countDistinct('PrimaryConditionGroup')).show()
df_claims.groupby('PrimaryConditionGroup').count().distinct().show(46)

+----------------------------+
|count(PrimaryConditionGroup)|
+----------------------------+
|                          46|
+----------------------------+

+---------------------+------+
|PrimaryConditionGroup| count|
+---------------------+------+
|                  NaN| 11410|
|              NEUMENT|171605|
|               RESPR4|138062|
|               INFEC4| 83552|
|               SEPSIS|   497|
|               STROKE|  8416|
|              PERVALV|  3518|
|               CANCRA|  5587|
|               CANCRB| 42895|
|              LIVERDZ|  2747|
|              SEIZURE| 20501|
|               CANCRM|  1096|
|               PRGNCY| 32004|
|               SKNAUT|107976|
|               HEART2| 54207|
|               HEART4| 28733|
|                  CHF| 13316|
|               RENAL1|   602|
|               RENAL2| 10922|
|               RENAL3| 52214|
|                ROAMI| 48821|
|               MISCL1|  4892|
|               MISCL5| 48307|
|             ARTHSPIN|288285|
|      

In [8]:
df_claims.agg(func.countDistinct('CharlsonIndex')).show()
df_claims.groupby('CharlsonIndex').count().distinct().show()

+--------------------+
|count(CharlsonIndex)|
+--------------------+
|                   4|
+--------------------+

+-------------+-------+
|CharlsonIndex|  count|
+-------------+-------+
|           5+|   5989|
|          3-4|  49479|
|            0|1356995|
|          1-2|1256527|
+-------------+-------+



#### All the values in the claims table are strings.  There are large numbers of null values across the claims dataframe.  In order to clean this up for use one would need to impute data for nulls (or remove altogether), recast any strings as integers, like length of stay or pay delay and update categorical field data to integerst (0, 1-2 becomes 1, 5+ becomes 5, etc).  Eventually, one will also one-hot encode the variables to transform categorical variables into boolean columns (i.e Gender becomes a column for Male and a column for Female with possible values of 0 and 1)

## Drug Data <a name="4.4Drug"></a>

[Back to 4.4](#Hughes)

In the Drug Count DataFrame (df_drug_count), what values can DrugCount take on? Are they all integers?

In [11]:
# Read in Drug Data
df_drug_count = load_csv_file('DrugCount.csv').cache()
# Check the data
print "Schema for drug_count DataFrame with %d rows:" %(df_drug_count.count())
df_drug_count.printSchema()
df_drug_count.show(1)

Schema for drug_count DataFrame with 818241 rows:
root
 |-- MemberID: string (nullable = true)
 |-- Year: string (nullable = true)
 |-- DSFS: string (nullable = true)
 |-- DrugCount: string (nullable = true)

+--------+----+-----------+---------+
|MemberID|Year|       DSFS|DrugCount|
+--------+----+-----------+---------+
|48925661|  Y2|9-10 months|       7+|
+--------+----+-----------+---------+
only showing top 1 row



In [14]:
df_drug_count.dtypes

[('MemberID', 'string'),
 ('Year', 'string'),
 ('DSFS', 'string'),
 ('DrugCount', 'string')]

In [12]:
df_drug_count.groupby('DrugCount').count().distinct().show()

+---------+------+
|DrugCount| count|
+---------+------+
|        1|263501|
|        2|188559|
|        3|129881|
|        4| 87783|
|        5| 57768|
|        6| 36731|
|       7+| 54018|
+---------+------+



In [15]:
columns_for_analysis = ['MemberID', 'Year', 'DSFS', 'DrugCount']
for column in columns_for_analysis:
    df_drug_count.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
|27379248|   21|
|67598133|   24|
|66711434|   32|
|36244478|   33|
|83078422|   18|
|27422357|    4|
|52820571|   18|
| 7988901|   35|
|65286285|   28|
|37508582|   23|
| 8760515|   23|
|85513602|   27|
|81777293|    7|
|70582587|   35|
|82986308|   13|
|14877553|   18|
| 7693430|   29|
|36315307|   18|
|69538151|   20|
|39501460|   31|
+--------+-----+
only showing top 20 rows

+----+------+
|Year| count|
+----+------+
|  Y1|281619|
|  Y2|276027|
|  Y3|260595|
+----+------+

+------------+------+
|        DSFS| count|
+------------+------+
| 8- 9 months| 62700|
|10-11 months| 46184|
| 7- 8 months| 67433|
| 6- 7 months| 69350|
|  0- 1 month|108980|
| 5- 6 months| 72274|
| 4- 5 months| 73325|
| 9-10 months| 55984|
| 3- 4 months| 77044|
| 2- 3 months| 78528|
| 1- 2 months| 81395|
|11-12 months| 25044|
+------------+------+

+---------+------+
|DrugCount| count|
+---------+------+
|        1|263501|
|        2|188559|
|        3|129881|


In [17]:
df_drug_count.describe(['MemberID']).show()

+-------+--------------------+
|summary|            MemberID|
+-------+--------------------+
|  count|              818241|
|   mean| 4.996952043978241E7|
| stddev|2.8939260548182726E7|
|    min|            10000665|
|    max|            99998627|
+-------+--------------------+



#### The Drug Count dataframe appears to be in much better shape than the claims dataframe.  There don't appear to be null entries.  Columns are structured as strings and would need to be recast, if needed.  One would also need to combine duplicate member IDs by Year while keeping track of the number of original records.  This can be done with sums, averages, min, max, etc

## Lab Data <a name="4.4Lab"></a>

[Back to 4.4](#Hughes)

In the Lab Count DataFrame (df_lab_count), what values can LabCount take on? Are they all integers?

In [18]:
sqlContext.clearCache()

In [19]:
# Read in Lab Data
df_lab_count = load_csv_file('LabCount.csv').cache()
# Check the data
print "Schema for lab_count DataFrame with %d rows:" %(df_lab_count.count())
df_lab_count.printSchema()
df_lab_count.show(1)

Schema for lab_count DataFrame with 361484 rows:
root
 |-- MemberID: string (nullable = true)
 |-- Year: string (nullable = true)
 |-- DSFS: string (nullable = true)
 |-- LabCount: string (nullable = true)

+--------+----+-----------+--------+
|MemberID|Year|       DSFS|LabCount|
+--------+----+-----------+--------+
|69258001|  Y3|2- 3 months|       1|
+--------+----+-----------+--------+
only showing top 1 row



In [20]:
df_lab_count.dtypes

[('MemberID', 'string'),
 ('Year', 'string'),
 ('DSFS', 'string'),
 ('LabCount', 'string')]

In [21]:
columns_for_analysis = ['MemberID', 'Year', 'DSFS', 'LabCount']
for column in columns_for_analysis:
    df_lab_count.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
| 9606836|    8|
|38282005|    9|
|55006914|    6|
|14991150|    8|
|50666100|    9|
| 2249261|   36|
| 7375082|    4|
|60552053|    6|
|17279194|   32|
| 2175588|    6|
|40216870|    4|
|  257428|    4|
|16891988|    3|
|24563556|    6|
| 6246086|    2|
| 3441743|    4|
|95385376|    8|
|88621960|    2|
|89761766|    4|
|27422357|    3|
+--------+-----+
only showing top 20 rows

+----+------+
|Year| count|
+----+------+
|  Y1|120162|
|  Y2|122416|
|  Y3|118906|
+----+------+

+------------+-----+
|        DSFS|count|
+------------+-----+
| 8- 9 months|23455|
|10-11 months|16054|
| 7- 8 months|24724|
| 6- 7 months|26103|
|  0- 1 month|97581|
| 5- 6 months|27460|
| 4- 5 months|26432|
| 9-10 months|20126|
| 3- 4 months|29750|
| 2- 3 months|30305|
| 1- 2 months|32279|
|11-12 months| 7215|
+------------+-----+

+--------+-----+
|LabCount|count|
+--------+-----+
|       1|93744|
|       2|54183|
|       3|43472|
|       4|38815|
|       5|3

In [22]:
df_lab_count.describe(['MemberID']).show()

+-------+--------------------+
|summary|            MemberID|
+-------+--------------------+
|  count|              361484|
|   mean|5.0017938519298226E7|
| stddev|2.8956121975431975E7|
|    min|            10000665|
|    max|            99998824|
+-------+--------------------+



#### The Lab Count dataframe appears to be in much better shape than the claims dataframe.  There don't appear to be null entries.  Columns are structured as strings and would need to be recast, if needed.  One would also need to combine duplicate member IDs by Year while keeping track of the number of original records.  This can be done with sums, averages, min, max, etc

## Member Data <a name="4.4Member"></a>

[Back to 4.4](#Hughes)

In the Member DataFrame, what values are present? Are they all integers?

In [24]:
# Read in Members Data
df_members = load_csv_file('Members.csv').cache()
# Check the data
print "Schema for members DataFrame with %d rows:" %(df_members.count())
df_members.printSchema()
df_members.show(1)

Schema for members DataFrame with 113000 rows:
root
 |-- MemberID: string (nullable = true)
 |-- AgeAtFirstClaim: string (nullable = true)
 |-- Sex: string (nullable = true)

+--------+---------------+---+
|MemberID|AgeAtFirstClaim|Sex|
+--------+---------------+---+
|14723353|          70-79|  M|
+--------+---------------+---+
only showing top 1 row



In [25]:
df_members.dtypes

[('MemberID', 'string'), ('AgeAtFirstClaim', 'string'), ('Sex', 'string')]

In [27]:
columns_for_analysis = ['MemberID', 'AgeAtFirstClaim', 'Sex']
for column in columns_for_analysis:
    df_members.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
|17279194|    1|
|90347313|    1|
| 5249057|    1|
|18032916|    1|
|51730545|    1|
|19787095|    1|
|13509475|    1|
|90062913|    1|
|93146634|    1|
|96110658|    1|
|78556616|    1|
|  788500|    1|
|89852530|    1|
|44486701|    1|
| 8760515|    1|
| 7028708|    1|
| 9462205|    1|
|54260751|    1|
|15761488|    1|
|78850646|    1|
+--------+-----+
only showing top 20 rows

+---------------+-----+
|AgeAtFirstClaim|count|
+---------------+-----+
|            NaN| 5753|
|          10-19|11319|
|          50-59|13329|
|          40-49|16111|
|            80+| 7621|
|            0-9|10791|
|          30-39|12435|
|          70-79|14514|
|          20-29| 8505|
|          60-69|12622|
+---------------+-----+

+---+-----+
|Sex|count|
+---+-----+
|NaN|17552|
|  F|51482|
|  M|43966|
+---+-----+



#### The Member Data is not as clean as the Drug or Lab data.  It has null values for gender and age at first claim.  I would recommend dropping the null values in this dataframe as they are not readily imputable.  One would also need to one-hot encode gender and age at first claim

## Target_Yr2 Data <a name="4.4Y2"></a>

[Back to 4.4](#Hughes)

In each of the Target DataFrames (df_target_Y2, df_target_Y3, df_target_Y4), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers?

In [29]:
# Read in Year 2 Target Variables
df_target_Y2 = load_csv_file('DaysInHospital_Y2.csv').cache()
# Check the data
print "Schema for target_Y2 DataFrame with %d rows:" %(df_target_Y2.count())
df_target_Y2.printSchema()
df_target_Y2.show(1)

Schema for target_Y2 DataFrame with 76038 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospital: string (nullable = true)

+--------+---------------+--------------+
|MemberID|ClaimsTruncated|DaysInHospital|
+--------+---------------+--------------+
|24027423|              0|             0|
+--------+---------------+--------------+
only showing top 1 row



In [31]:
df_target_Y2.dtypes

[('MemberID', 'string'),
 ('ClaimsTruncated', 'string'),
 ('DaysInHospital', 'string')]

In [33]:
columns_for_analysis = ['MemberID', 'ClaimsTruncated', 'DaysInHospital']
for column in columns_for_analysis:
    df_target_Y2.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
|28808907|    1|
|27346839|    1|
|39501460|    1|
| 7375082|    1|
| 8738411|    1|
|28800833|    1|
|79042057|    1|
|78922204|    1|
|66625953|    1|
|50893522|    1|
|73327222|    1|
|84915601|    1|
|22079844|    1|
|21166770|    1|
|66948270|    1|
|59418162|    1|
| 1487835|    1|
|47684717|    1|
|17741947|    1|
|17646448|    1|
+--------+-----+
only showing top 20 rows

+---------------+-----+
|ClaimsTruncated|count|
+---------------+-----+
|              0|72067|
|              1| 3971|
+---------------+-----+

+--------------+-----+
|DaysInHospital|count|
+--------------+-----+
|             0|64269|
|             1| 4835|
|             2| 2366|
|             3| 1453|
|             4|  977|
|             5|  565|
|             6|  373|
|             7|  256|
|             8|  173|
|             9|  148|
|            10|  106|
|            11|   80|
|            12|   73|
|            13|   61|
|            14|   42|
|      

## Target_Yr3 Data <a name="4.4Y3"></a>

[Back to 4.4](#Hughes)

In each of the Target DataFrames (df_target_Y2, df_target_Y3, df_target_Y4), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers?

In [34]:
# Read in Year 3 Target Variables
df_target_Y3 = load_csv_file('DaysInHospital_Y3.csv').cache()
# Check the data
print "Schema for target_Y3 DataFrame with %d rows:" %(df_target_Y3.count())
df_target_Y3.printSchema()
df_target_Y3.show(1)

Schema for target_Y3 DataFrame with 71435 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospital: string (nullable = true)

+--------+---------------+--------------+
|MemberID|ClaimsTruncated|DaysInHospital|
+--------+---------------+--------------+
|90963501|              0|             0|
+--------+---------------+--------------+
only showing top 1 row



In [35]:
df_target_Y3.dtypes

[('MemberID', 'string'),
 ('ClaimsTruncated', 'string'),
 ('DaysInHospital', 'string')]

In [36]:
for column in columns_for_analysis:
    df_target_Y3.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
|18927481|    1|
|30238844|    1|
|49057069|    1|
|43332540|    1|
|58660696|    1|
|15761488|    1|
|48590284|    1|
| 8954484|    1|
|42558734|    1|
| 7375082|    1|
|89861396|    1|
|21687169|    1|
|81284795|    1|
| 8159324|    1|
|36244478|    1|
|81952531|    1|
|35315443|    1|
|98711556|    1|
|76756557|    1|
|73644635|    1|
+--------+-----+
only showing top 20 rows

+---------------+-----+
|ClaimsTruncated|count|
+---------------+-----+
|              0|67391|
|              1| 4044|
+---------------+-----+

+--------------+-----+
|DaysInHospital|count|
+--------------+-----+
|             0|60706|
|             1| 4464|
|             2| 2182|
|             3| 1429|
|             4|  842|
|             5|  528|
|             6|  287|
|             7|  218|
|             8|  143|
|             9|  115|
|            10|  103|
|            11|   65|
|            12|   62|
|            13|   50|
|            14|   23|
|      

## Target_Yr4 Data <a name="4.4Y4"></a>

[Back to 4.4](#Hughes)

In each of the Target DataFrames (df_target_Y2, df_target_Y3, df_target_Y4), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers?

In [38]:
# Read in Year 4 Target Variables
df_target_Y4 = load_csv_file('Target.csv').cache()
# Check the data
print "Schema for target_Y4 DataFrame with %d rows:" %(df_target_Y4.count())
df_target_Y4.printSchema()
df_target_Y4.show(1)

Schema for target_Y4 DataFrame with 70942 rows:
root
 |-- MemberID: string (nullable = true)
 |-- ClaimsTruncated: string (nullable = true)
 |-- DaysInHospital: string (nullable = true)

+--------+---------------+--------------+
|MemberID|ClaimsTruncated|DaysInHospital|
+--------+---------------+--------------+
|20820036|              0|           NaN|
+--------+---------------+--------------+
only showing top 1 row



In [39]:
df_target_Y4.dtypes

[('MemberID', 'string'),
 ('ClaimsTruncated', 'string'),
 ('DaysInHospital', 'string')]

In [40]:
for column in columns_for_analysis:
    df_target_Y4.groupBy(column).count().show()

+--------+-----+
|MemberID|count|
+--------+-----+
|56778797|    1|
|79455022|    1|
|28757067|    1|
|10100129|    1|
|54636033|    1|
|81617254|    1|
|47970610|    1|
|62223236|    1|
|22950189|    1|
|58943638|    1|
|14877553|    1|
|36315307|    1|
| 8159324|    1|
|33683928|    1|
|49226403|    1|
| 7693430|    1|
|29762321|    1|
|40216870|    1|
|27346839|    1|
| 8738411|    1|
+--------+-----+
only showing top 20 rows

+---------------+-----+
|ClaimsTruncated|count|
+---------------+-----+
|              0|66974|
|              1| 3968|
+---------------+-----+

+--------------+-----+
|DaysInHospital|count|
+--------------+-----+
|           NaN|70942|
+--------------+-----+



Days in Hospital for Yr2 and Yr3 are strings from 0 to 15.  These will need to be converted to integer.  The values for Days in Hospital is null for Yr4 because this is the target value.  Claims Truncated take on the value 0 or 1.  Currently they are string values.  They will need to be converted to integer